topology/pod-cidr/src/main/java/org/apache/hadoop/net/PodCIDRToNodeMapping.java

```diff
@@ -0,0 +1,307 @@
+    if (podCIDRLookup != null) {
+      // TODO. Support refresh.
+      return podCIDRLookup;
+    }
+    podCIDRLookup = PodCIDRLookup.fetchPodCIDR(getOrCreateKubernetesClient());
+    if (log.isDebugEnabled()) {
+      log.debug("Fetched pod CIDR per node and built a lookup" + podCIDRLookup);
+    }
+    return podCIDRLookup;
+  }
+
+  private synchronized KubernetesClient getOrCreateKubernetesClient() {
+    if (kubernetesClient != null) {
+      return kubernetesClient;
+    }
+    // Disable the ping thread that is not daemon, in order to allow the main thread to shut down
+    // upon errors. Otherwise, the namenode will hang indefinitely.
+    Config config = new ConfigBuilder()
+        .withWebsocketPingInterval(0)
+        .build();
+    // Use a Dispatcher with a custom executor service that creates daemon threads. The default
+    // executor service used by Dispatcher creates non-daemon threads.
+    OkHttpClient httpClient = HttpClientUtils.createHttpClient(config).newBuilder()
+        .dispatcher(new Dispatcher(
+            Executors.newCachedThreadPool(
+                new ThreadFactoryBuilder().setDaemon(true)
+                    .setNameFormat("k8s-topology-plugin-%d")
+                    .build())))
+        .build();
+    kubernetesClient = new DefaultKubernetesClient(httpClient, config);
+    return kubernetesClient;
+  }
+
+  private static class PodCIDRLookup {
+
+    // K8s cluster node names.
+    private final Set<String> nodeNames;
+    // K8s cluster node names indexed by pod subnet information. The top level map contains
+    // netmask strings as keys. The second level map contains network addresses as keys.
+    private final Map<String, Map<String, String>> nodeNameBySubnet;
+
+    PodCIDRLookup() {
+      this(Collections.<String>emptySet(), Collections.<String, Map<String, String>>emptyMap());
+    }
+
+    private PodCIDRLookup(Set<String> nodeNames,
+        Map<String, Map<String, String>> nodeNameBySubnet) {
+      this.nodeNames = nodeNames;
+      this.nodeNameBySubnet = nodeNameBySubnet;
+    }
+
+    boolean containsNode(String nodeName) {
+      return nodeNames.contains(nodeName);
+    }
+
+    String findNodeByPodIP(String podIP) {
+      for (Map.Entry<String, Map<String, String>> entry : nodeNameBySubnet.entrySet()) {
+        String netmask = entry.getKey();
+        SubnetInfo subnetInfo;
+        try {
+          subnetInfo = new SubnetUtils(podIP, netmask).getInfo();
+        } catch (IllegalArgumentException e) {
+          log.warn(e);
+          continue;
+        }
+        String networkAddress = subnetInfo.getNetworkAddress();
+        Map<String, String> nodeNameByNetworkAddress = entry.getValue();
+        String nodeName = nodeNameByNetworkAddress.get(networkAddress);
+        if (nodeName != null) {
+          return nodeName;
+        }
+      }
+      return "";
+    }
+
+    static PodCIDRLookup fetchPodCIDR(KubernetesClient kubernetesClient) {
```

Comment:
Could you add comment with an example and what the inputs and conversions would look like.

optionally: also consider breaking this up and possibly convert the network address/nodename into a separate class.